README.md
# Circlator
A tool to circularize genome assemblies.

[![Build Status](https://travis-ci.org/sanger-pathogens/circlator.svg?branch=master)](https://travis-ci.org/sanger-pathogens/circlator)   
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-brightgreen.svg)](https://github.com/sanger-pathogens/circlator/blob/master/LICENSE)   
[![status](https://img.shields.io/badge/GenomeBiology-10.1186-brightgreen.svg)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0849-0)  
[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg)](http://bioconda.github.io/recipes/circlator/README.html)  
[![Container ready](https://img.shields.io/badge/container-ready-brightgreen.svg)](https://quay.io/repository/biocontainers/circlator)  
[![Docker Build Status](https://img.shields.io/docker/build/sangerpathogens/circlator.svg)](https://hub.docker.com/r/sangerpathogens/circlator)  
[![Docker Pulls](https://img.shields.io/docker/pulls/sangerpathogens/circlator.svg)](https://hub.docker.com/r/sangerpathogens/circlator)  
[![codecov](https://codecov.io/gh/sanger-pathogens/circlator/branch/master/graph/badge.svg)](https://codecov.io/gh/sanger-pathogens/circlator)

## Contents
  * [Introduction](#introduction)
  * [Installation](#installation)
  * [Usage](#usage)
  * [License](#license)
  * [Feedback/Issues](#feedbackissues)
  * [Citation](#citation)
  * [Further Information](#further-information)

## Introduction
Circlator will attempt to identify each circular sequence and output a linearised version of it. It does this by assembling all reads that map to contig ends and comparing the resulting contigs with the input assembly.

## Installation
For how to install Circlator, please see the [Circlator website](http://sanger-pathogens.github.io/circlator/).

## Usage
```
Usage: circlator <command> [options] <required arguments>

To get minimal usage for a command use:
circlator command

To get full help for a command use one of:
circlator command -h
circlator command --help


Available commands:

all        Run mapreads, bam2reads, assemble, merge, clean, fixstart
mapreads   Map reads to assembly
bam2reads  Make reads from mapping to be reassembled
assemble   Run assembly using reads from bam2reads
merge      Merge original assembly and new assembly made by assemble
clean      Remove small and completely contained contigs from assembly
fixstart   Change start position of circular sequences
minimus2   Run the minimus2 based circularisation pipeline
get_dnaa   Download file of dnaA (or other of user's choice) genes
progcheck  Checks dependencies are installed
test       Run Circlator on a small test set
version    Print version and exit
```
For more information on how to use Circlator, please see the [Circlator wiki page](https://github.com/sanger-pathogens/circlator/wiki).

## License
Circlator is free software, licensed under [GPLv3](https://github.com/sanger-pathogens/circlator/blob/master/LICENSE).

## Feedback/Issues
Please report any issues to the [issues page](https://github.com/sanger-pathogens/circlator/issues) or email circlator-help@sanger.ac.uk.

## Citation
If you use this software please cite:

__Circlator: automated circularization of genome assemblies using long sequencing reads__   
Hunt et al, Genome Biology 2015 Dec 29;16(1):294. doi: [10.1186/s13059-015-0849-0](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0849-0)

__[BWA](http://arxiv.org/abs/1303.3997):__ Li, H et al. Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM. arXiv:1303.3997.  

__[MUMmer](http://genomebiology.com/content/5/2/R12):__ Kurtz, S. et al. Versatile and open software for comparing large genomes. Genome Biol. 5, R12 (2004).  

__[Prodigal](http://www.biomedcentral.com/1471-2105/11/119):__ Hyatt, D. et al. Prodigal: prokaryotic gene recognition and translation initiation site identification. BMC Bioinformatics 11, 119 (2010).  

__[SAMtools](http://bioinformatics.oxfordjournals.org/content/25/16/2078.abstract):__ Li, H. et al. The Sequence Alignment/Map format and SAMtools. Bioinformatics 25, 2078–9 (2009).  

__[SPAdes](http://online.liebertpub.com/doi/abs/10.1089/cmb.2012.0021)__: Bankevich, A. et al. SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing. J. Comput. Biol. 19, 455–77 (2012)  

## Further Information
For more information on this software see:
* the [Circlator website](http://sanger-pathogens.github.io/circlator/)
* the [Circlator wiki page](https://github.com/sanger-pathogens/circlator/wiki)

CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/sanger-pathogens/circlator/tree/HEAD)

[Full Changelog](https://github.com/sanger-pathogens/circlator/compare/v1.5.5...HEAD)

**Fixed bugs:**

- Infinite recursion in clean.py ? [\#123](https://github.com/sanger-pathogens/circlator/issues/123)

**Closed issues:**

- how to use circlator on computer without internet access [\#108](https://github.com/sanger-pathogens/circlator/issues/108)

**Merged pull requests:**

- fix readme and add to codecov [\#141](https://github.com/sanger-pathogens/circlator/pull/141) ([ssjunnebo](https://github.com/ssjunnebo))
- Include Docker  [\#137](https://github.com/sanger-pathogens/circlator/pull/137) ([ssjunnebo](https://github.com/ssjunnebo))
- edit Dockerfile [\#136](https://github.com/sanger-pathogens/circlator/pull/136) ([ssjunnebo](https://github.com/ssjunnebo))
- add Dockerfile [\#135](https://github.com/sanger-pathogens/circlator/pull/135) ([ssjunnebo](https://github.com/ssjunnebo))

## [v1.5.5](https://github.com/sanger-pathogens/circlator/tree/v1.5.5) (2018-01-31)
[Full Changelog](https://github.com/sanger-pathogens/circlator/compare/v1.5.4...v1.5.5)

**Closed issues:**

- nucmer version detected by "circlator progcheck" but not "circlator all" [\#122](https://github.com/sanger-pathogens/circlator/issues/122)
- Canu version is not being detected properly [\#113](https://github.com/sanger-pathogens/circlator/issues/113)

**Merged pull requests:**

- failsafe to prevent infinite recursion [\#126](https://github.com/sanger-pathogens/circlator/pull/126) ([pathdevg](https://github.com/pathdevg))

## [v1.5.4](https://github.com/sanger-pathogens/circlator/tree/v1.5.4) (2018-01-31)
[Full Changelog](https://github.com/sanger-pathogens/circlator/compare/v1.5.3...v1.5.4)

**Closed issues:**

- Problems running progcheck, error with installation? [\#124](https://github.com/sanger-pathogens/circlator/issues/124)

**Merged pull requests:**

- Add tests for external version string checking [\#125](https://github.com/sanger-pathogens/circlator/pull/125) ([andrewjpage](https://github.com/andrewjpage))

## [v1.5.3](https://github.com/sanger-pathogens/circlator/tree/v1.5.3) (2017-11-09)
[Full Changelog](https://github.com/sanger-pathogens/circlator/compare/v1.5.1...v1.5.3)

**Closed issues:**

- Clarification regarding nanopore data [\#119](https://github.com/sanger-pathogens/circlator/issues/119)
- nucmer version isn't detected correctly [\#117](https://github.com/sanger-pathogens/circlator/issues/117)
- Error when running circlator: can't find canu version [\#111](https://github.com/sanger-pathogens/circlator/issues/111)
- progcheck: found nucmer but couldn't get version [\#110](https://github.com/sanger-pathogens/circlator/issues/110)
- Installation problem: command not found after pip3 install circlator [\#109](https://github.com/sanger-pathogens/circlator/issues/109)
- memory usage [\#106](https://github.com/sanger-pathogens/circlator/issues/106)
- Recompile with -fPIC error git / pip3 installs [\#103](https://github.com/sanger-pathogens/circlator/issues/103)
- fixstart with user supplied gene sequences does not work [\#93](https://github.com/sanger-pathogens/circlator/issues/93)

**Merged pull requests:**

- version bump [\#120](https://github.com/sanger-pathogens/circlator/pull/120) ([ssjunnebo](https://github.com/ssjunnebo))
- fixes \#110, \#117 [\#118](https://github.com/sanger-pathogens/circlator/pull/118) ([spock](https://github.com/spock))
- circlator version [\#112](https://github.com/sanger-pathogens/circlator/pull/112) ([ssjunnebo](https://github.com/ssjunnebo))
- update AUTHORS and LICENSE [\#94](https://github.com/sanger-pathogens/circlator/pull/94) ([ssjunnebo](https://github.com/ssjunnebo))

## [v1.5.1](https://github.com/sanger-pathogens/circlator/tree/v1.5.1) (2017-04-21)
[Full Changelog](https://github.com/sanger-pathogens/circlator/compare/v1.5.0...v1.5.1)

**Closed issues:**

- Circlator with short reads? [\#88](https://github.com/sanger-pathogens/circlator/issues/88)

**Merged pull requests:**

- Canu1.5 [\#89](https://github.com/sanger-pathogens/circlator/pull/89) ([martinghunt](https://github.com/martinghunt))

## [v1.5.0](https://github.com/sanger-pathogens/circlator/tree/v1.5.0) (2017-04-03)
[Full Changelog](https://github.com/sanger-pathogens/circlator/compare/v1.4.1...v1.5.0)

**Closed issues:**

- version of SPAdes ? [\#85](https://github.com/sanger-pathogens/circlator/issues/85)
- It looks like circlator keeps a common k-mer at the ends of a contig after changing start position. What for? [\#82](https://github.com/sanger-pathogens/circlator/issues/82)
- Error running circlator with option  --assemble\_not\_only\_assembler [\#80](https://github.com/sanger-pathogens/circlator/issues/80)

**Merged pull requests:**

- Add canu [\#87](https://github.com/sanger-pathogens/circlator/pull/87) ([martinghunt](https://github.com/martinghunt))
- Included option --CanuCorrectedErrorRate to allow user to modify opti… [\#86](https://github.com/sanger-pathogens/circlator/pull/86) ([NicolaDM](https://github.com/NicolaDM))
- Splitting reads for small contigs when using Canu [\#84](https://github.com/sanger-pathogens/circlator/pull/84) ([NicolaDM](https://github.com/NicolaDM))
- Adding the option to use Canu instead of SPAdes [\#83](https://github.com/sanger-pathogens/circlator/pull/83) ([NicolaDM](https://github.com/NicolaDM))

## [v1.4.1](https://github.com/sanger-pathogens/circlator/tree/v1.4.1) (2017-01-24)
[Full Changelog](https://github.com/sanger-pathogens/circlator/compare/v1.4.0...v1.4.1)

**Closed issues:**

- Circularization of BAC contigs [\#79](https://github.com/sanger-pathogens/circlator/issues/79)
- failure to circularise: cannot use this pair because longer match was found [\#75](https://github.com/sanger-pathogens/circlator/issues/75)
- samtools command error [\#66](https://github.com/sanger-pathogens/circlator/issues/66)

**Merged pull requests:**


circlator/assemble.py
import os
import tempfile
import sys
import shutil
import pyfastaq
from circlator import common, external_progs

class Error (Exception): pass

class Assembler:
    def __init__(self,
      reads,
      outdir,
      threads=1,
      spades_kmers=None,
      careful=True,
      only_assembler=True,
      verbose=False,
      spades_use_first_success=False,
      assembler='spades',
      genomeSize=100000, # only matters for Canu if correcting reads (which we're not)
      data_type='pacbio-corrected',
    ):
        self.outdir = os.path.abspath(outdir)
        self.reads = os.path.abspath(reads)
        if not os.path.exists(self.reads):
            raise Error('Reads file not found:' + self.reads)

        self.verbose = verbose
        self.samtools = external_progs.make_and_check_prog('samtools', verbose=self.verbose)
        self.threads = threads
        self.assembler = assembler

        if self.assembler == 'spades':
            self.spades = external_progs.make_and_check_prog('spades', verbose=self.verbose, required=True)
            self.spades_kmers = self._build_spades_kmers(spades_kmers)
            self.spades_use_first_success = spades_use_first_success
            self.careful = careful
            self.only_assembler = only_assembler
        elif self.assembler == 'canu':
            self.canu = external_progs.make_and_check_prog('canu', verbose=self.verbose, required=True)
            self.genomeSize=genomeSize
            self.data_type = data_type
        else:
            raise Error('Unknown assembler: "' + self.assembler + '". cannot continue')



    def _build_spades_kmers(self, kmers):
        if kmers is None:
            return [127,117,107,97,87,77]
        elif type(kmers) == str:
            try:
                kmer_list = [int(k) for k in kmers.split(',')]
            except:
                raise Error('Error getting list of kmers from:' + str(kmers))
            return kmer_list
        elif type(kmers) == list:
            return kmers
        else:
            raise Error('Error getting list of kmers from:' + str(kmers))


    def _make_spades_command(self, kmer, outdir):
        cmd = [
            self.spades.exe(),
            '-s', self.reads,
            '-o', outdir,
            '-t', str(self.threads),
            '-k', str(kmer),
        ]

        if self.careful:
            cmd.append('--careful')

        if self.only_assembler:
            cmd.append('--only-assembler')

        return ' '.join(cmd)


    def _make_canu_command(self, outdir, out_name):
        cmd = [
            self.canu.exe(),
            '-useGrid=false',
            'gnuplotTested=true',
            '-assemble',
            'genomeSize='+str(float(self.genomeSize)/1000000)+'m',
            '-d', outdir,
            '-p', out_name,
            '-'+self.data_type,
            self.reads,
        ]
        return ' '.join(cmd)


    def run_spades_once(self, kmer, outdir):
        cmd = self._make_spades_command(kmer, outdir)
        return common.syscall(cmd, verbose=self.verbose, allow_fail=True)


    def run_spades(self, stop_at_first_success=False):
        '''Runs spades on all kmers. Each a separate run because SPAdes dies if any kmer does
           not work. Chooses the 'best' assembly to be the one with the biggest N50'''
        n50 = {}
        kmer_to_dir = {}

        for k in self.spades_kmers:
            tmpdir = tempfile.mkdtemp(prefix=self.outdir + '.tmp.spades.' + str(k) + '.', dir=os.getcwd())
            kmer_to_dir[k] = tmpdir
            ok, errs = self.run_spades_once(k, tmpdir)
            if ok:
                contigs_fasta = os.path.join(tmpdir, 'contigs.fasta')
                contigs_fai = contigs_fasta + '.fai'
                common.syscall(self.samtools.exe() + ' faidx ' + contigs_fasta, verbose=self.verbose)
                stats = pyfastaq.tasks.stats_from_fai(contigs_fai)
                if stats['N50'] != 0:
                    n50[k] = stats['N50']

                    if stop_at_first_success:
                        break

        if len(n50) > 0:
            if self.verbose:
                print('[assemble]\tkmer\tN50')
                for k in sorted(n50):
                    print('[assemble]', k, n50[k], sep='\t')

            best_k = None

            for k in sorted(n50):
                if best_k is None or n50[k] >= n50[best_k]:
                    best_k = k

            assert best_k is not None

            for k, directory in kmer_to_dir.items():
                if k == best_k:
                    if self.verbose:
                        print('[assemble] using assembly with kmer', k)
                    os.rename(directory, self.outdir)
                else:
                    shutil.rmtree(directory)
        else:
            raise Error('Error running SPAdes. Output directories are:\n  ' + '\n  '.join(kmer_to_dir.values()) + '\nThe reason why should be in the spades.log file in each directory.')


    @classmethod
    def _rename_canu_contigs(cls, infile, outfile):
        with open (infile) as f_in:
            with open(outfile, 'w') as f_out:
                for line in f_in:
                    if line.startswith('>'):
                        print(line.split()[0], file=f_out)
                    else:
                        print(line, end='', file=f_out)


    def run_canu(self):
        '''Runs canu instead of spades'''
        cmd = self._make_canu_command(self.outdir,'canu')
        ok, errs = common.syscall(cmd, verbose=self.verbose, allow_fail=False)
        if not ok:
            raise Error('Error running Canu.')

        original_contigs = os.path.join(self.outdir, 'canu.contigs.fasta')
        renamed_contigs = os.path.join(self.outdir, 'contigs.fasta')
        Assembler._rename_canu_contigs(original_contigs, renamed_contigs)
        original_gfa = os.path.join(self.outdir, 'canu.contigs.gfa')
        renamed_gfa = os.path.join(self.outdir, 'contigs.gfa')
        os.rename(original_gfa, renamed_gfa)


    def run(self):
        if self.assembler == 'spades':
            self.run_spades(stop_at_first_success=self.spades_use_first_success)
        elif self.assembler == 'canu':
            self.run_canu()
        else:
            raise Error('Unknown assembler: "' + self.assembler + '". cannot continue')
